x-pack/legacy/plugins/infra/server/lib/adapters/metadata/elasticsearch_metadata_adapter.ts

```diff
@@ -73,6 +82,134 @@ export class ElasticsearchMetadataAdapter implements InfraMetadataAdapter {
       buckets,
     };
   }
+
+  public async getCloudMetricMetadata(
+    req: InfraFrameworkRequest,
+    sourceConfiguration: InfraSourceConfiguration,
+    instanceId: string
+  ): Promise<InfraCloudMetricsAdapterResponse> {
+    const metricQuery = {
+      allowNoIndices: true,
+      ignoreUnavailable: true,
+      index: sourceConfiguration.metricAlias,
+      body: {
+        query: {
+          bool: {
+            filter: [{ match: { 'cloud.instance.id': instanceId } }],
+            should: CLOUD_METRICS_MODULES.map(module => ({ match: { 'event.module': module } })),
+          },
+        },
+        size: 0,
+        aggs: {
+          metrics: {
+            terms: {
+              field: 'event.dataset',
+              size: 1000,
+            },
+          },
+        },
+      },
+    };
+
+    const response = await this.framework.callWithRequest<
+      {},
+      {
+        metrics?: InfraMetadataAggregationResponse;
+      }
+    >(req, 'search', metricQuery);
+
+    const buckets =
+      response.aggregations && response.aggregations.metrics
+        ? response.aggregations.metrics.buckets
+        : [];
+
+    return { buckets };
+  }
+
+  public async getNodeInfo(
+    req: InfraFrameworkRequest,
+    sourceConfiguration: InfraSourceConfiguration,
+    nodeId: string,
+    nodeType: InfraNodeType
+  ): Promise<InfraNodeInfo> {
+    // If the nodeType is a Kubernetes pod then we need to get the node info
+    // from a host record instead of a pod. This is due to the fact that any host
+    // can report pod details and we can't rely on the host/cloud information associated
+    // with the kubernetes.pod.uid. We need to first lookup the `kubernetes.node.name`
+    // then use that to lookup the host's node information.
```

Comment:
Given this, is it even correct that we use `kubernetes.pod.uid` as the main node identifier for Pod nodes?

Reply:
`kubernetes.pod.uid` is the right field, the issue is the metadata attached to the documents comes from the "reporting" Metricbeat process. It's similar to how AWS metrics are collected, each Kubernetes Node can report details about the cluster. If anything I think the question is "Should we be adding the host/cloud metadata to Pods and Containers?"

Reply:
Thanks for the explanation. Yes, I think we should add the host/cloud metadata to pods and containers.